{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genic Notation"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "According to the paper, they have mapped to the human reference genome (hg19).\n",
    "HG19 is the alias to GRCh37."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Human Gene Annotation - GENCODE Release 19 (GRCh37.p13)\n",
    "\n",
    "https://www.gencodegenes.org/human/release_19.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import gzip\n",
    "import shutil\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# File_in contains the comprehensive gene annotation on the reference chromosomes only\n",
    "file_in = 'Data/ComprehensiveGeneAnnotation/gencode.v19.annotation.gff3.gz'\n",
    "file_out = 'Data/ComprehensiveGeneAnnotation/gencode.v19.annotation.gff3'\n",
    "\n",
    "# # File_in contains the basic gene annotation on the reference chromosomes only\n",
    "# # This is a subset of the corresponding comprehensive annotation, including only those transcripts tagged as 'basic' in every gene\n",
    "# # This is the main annotation file for most users\n",
    "# file_in = 'gencode.v42.basic.annotation.gff3.gz'\n",
    "# file_out = 'gencode.v42.basic.annotation.gff3'\n",
    "\n",
    "with gzip.open(file_in, 'rb') as f_in:\n",
    "    with open(file_out, 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read dataset\n",
    "data_df = pd.read_csv(file_out, sep='\\t', header=None, skiprows=7)\n",
    "#data_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2615590, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.4;gene_id=ENSG00000223972.4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.4;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1           2        3        4  5  6  7  \\\n",
       "0  chr1  HAVANA        gene  11869.0  14412.0  .  +  .   \n",
       "1  chr1  HAVANA  transcript  11869.0  14409.0  .  +  .   \n",
       "2  chr1  HAVANA        exon  11869.0  12227.0  .  +  .   \n",
       "3  chr1  HAVANA        exon  12613.0  12721.0  .  +  .   \n",
       "4  chr1  HAVANA        exon  13221.0  14409.0  .  +  .   \n",
       "\n",
       "                                                   8  \n",
       "0  ID=ENSG00000223972.4;gene_id=ENSG00000223972.4...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.4;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(data_df.shape)    # 2.6M x 9\n",
    "data_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy dataset\n",
    "data_df_mod = data_df.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2615566, 9)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove rows that contain comments\n",
    "\n",
    "# data_df_mod.shape\n",
    "# data_df_mod.isnull().values.any()\n",
    "# data_df_mod.isnull().sum()\n",
    "# print(data_df_mod[data_df_mod.isnull().any(axis=1)])\n",
    "data_df_mod = data_df_mod.dropna(axis = 0, how = 'any')\n",
    "# print(data_df_mod[data_df_mod.isnull().any(axis=1)])\n",
    "# data_df_mod.shape\n",
    "data_df_mod.shape   # 2.6M x 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create column ID\n",
    "\n",
    "data_df_mod['ID'] = data_df_mod.apply(\n",
    "    lambda x: x[8][x[8].find('ID=')+3:x[8].find('.')], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.4;gene_id=ENSG00000223972.4...</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.4;...</td>\n",
       "      <td>ENST00000456328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1           2        3        4  5  6  7  \\\n",
       "0  chr1  HAVANA        gene  11869.0  14412.0  .  +  .   \n",
       "1  chr1  HAVANA  transcript  11869.0  14409.0  .  +  .   \n",
       "2  chr1  HAVANA        exon  11869.0  12227.0  .  +  .   \n",
       "3  chr1  HAVANA        exon  12613.0  12721.0  .  +  .   \n",
       "4  chr1  HAVANA        exon  13221.0  14409.0  .  +  .   \n",
       "\n",
       "                                                   8                    ID  \n",
       "0  ID=ENSG00000223972.4;gene_id=ENSG00000223972.4...       ENSG00000223972  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.4;...       ENST00000456328  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  exon:ENST00000456328  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  exon:ENST00000456328  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  exon:ENST00000456328  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df_mod.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create column Parent\n",
    "\n",
    "#The find() method returns the index of first occurrence of the substring (if found). If not found, it returns -1.\n",
    "data_df_mod['Parent'] = data_df_mod.apply(\n",
    "    lambda x: x[8].split(';')[1][7:x[8].split(';')[1].find('.')] if x[2] != 'gene' else 'nan', \n",
    "    axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.4;gene_id=ENSG00000223972.4...</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.4;...</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11872.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000515242.2;Parent=ENSG00000223972.4;...</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>11872.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000515242.2:1;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000515242</td>\n",
       "      <td>ENST00000515242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000515242.2:2;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000515242</td>\n",
       "      <td>ENST00000515242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>13225.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000515242.2:3;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000515242</td>\n",
       "      <td>ENST00000515242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000518655.2;Parent=ENSG00000223972.4;...</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1           2        3        4  5  6  7  \\\n",
       "0  chr1   HAVANA        gene  11869.0  14412.0  .  +  .   \n",
       "1  chr1   HAVANA  transcript  11869.0  14409.0  .  +  .   \n",
       "2  chr1   HAVANA        exon  11869.0  12227.0  .  +  .   \n",
       "3  chr1   HAVANA        exon  12613.0  12721.0  .  +  .   \n",
       "4  chr1   HAVANA        exon  13221.0  14409.0  .  +  .   \n",
       "5  chr1  ENSEMBL  transcript  11872.0  14412.0  .  +  .   \n",
       "6  chr1  ENSEMBL        exon  11872.0  12227.0  .  +  .   \n",
       "7  chr1  ENSEMBL        exon  12613.0  12721.0  .  +  .   \n",
       "8  chr1  ENSEMBL        exon  13225.0  14412.0  .  +  .   \n",
       "9  chr1  ENSEMBL  transcript  11874.0  14409.0  .  +  .   \n",
       "\n",
       "                                                   8                    ID  \\\n",
       "0  ID=ENSG00000223972.4;gene_id=ENSG00000223972.4...       ENSG00000223972   \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.4;...       ENST00000456328   \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  exon:ENST00000456328   \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  exon:ENST00000456328   \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  exon:ENST00000456328   \n",
       "5  ID=ENST00000515242.2;Parent=ENSG00000223972.4;...       ENST00000515242   \n",
       "6  ID=exon:ENST00000515242.2:1;Parent=ENST0000051...  exon:ENST00000515242   \n",
       "7  ID=exon:ENST00000515242.2:2;Parent=ENST0000051...  exon:ENST00000515242   \n",
       "8  ID=exon:ENST00000515242.2:3;Parent=ENST0000051...  exon:ENST00000515242   \n",
       "9  ID=ENST00000518655.2;Parent=ENSG00000223972.4;...       ENST00000518655   \n",
       "\n",
       "            Parent  \n",
       "0              nan  \n",
       "1  ENSG00000223972  \n",
       "2  ENST00000456328  \n",
       "3  ENST00000456328  \n",
       "4  ENST00000456328  \n",
       "5  ENSG00000223972  \n",
       "6  ENST00000515242  \n",
       "7  ENST00000515242  \n",
       "8  ENST00000515242  \n",
       "9  ENSG00000223972  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df_mod.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create column with Exon length\n",
    "data_df_mod['ExonLength'] = data_df_mod.apply(\n",
    "    lambda x: x[4]-x[3] if x[2]=='exon' else 0,\n",
    "    axis = 1\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>ExonLength</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.4;gene_id=ENSG00000223972.4...</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.4;...</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>358.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>1188.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11872.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000515242.2;Parent=ENSG00000223972.4;...</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>11872.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000515242.2:1;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000515242</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>355.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000515242.2:2;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000515242</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>13225.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000515242.2:3;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000515242</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>1187.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000518655.2;Parent=ENSG00000223972.4;...</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000518655.2:1;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000518655</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>353.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12595.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000518655.2:2;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000518655</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>126.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>13403.0</td>\n",
       "      <td>13655.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000518655.2:3;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000518655</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>252.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>13661.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000518655.2:4;Parent=ENST0000051...</td>\n",
       "      <td>exon:ENST00000518655</td>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>748.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>13670.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000450305.2;Parent=ENSG00000223972.4;...</td>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0        1           2        3        4  5  6  7  \\\n",
       "0   chr1   HAVANA        gene  11869.0  14412.0  .  +  .   \n",
       "1   chr1   HAVANA  transcript  11869.0  14409.0  .  +  .   \n",
       "2   chr1   HAVANA        exon  11869.0  12227.0  .  +  .   \n",
       "3   chr1   HAVANA        exon  12613.0  12721.0  .  +  .   \n",
       "4   chr1   HAVANA        exon  13221.0  14409.0  .  +  .   \n",
       "5   chr1  ENSEMBL  transcript  11872.0  14412.0  .  +  .   \n",
       "6   chr1  ENSEMBL        exon  11872.0  12227.0  .  +  .   \n",
       "7   chr1  ENSEMBL        exon  12613.0  12721.0  .  +  .   \n",
       "8   chr1  ENSEMBL        exon  13225.0  14412.0  .  +  .   \n",
       "9   chr1  ENSEMBL  transcript  11874.0  14409.0  .  +  .   \n",
       "10  chr1  ENSEMBL        exon  11874.0  12227.0  .  +  .   \n",
       "11  chr1  ENSEMBL        exon  12595.0  12721.0  .  +  .   \n",
       "12  chr1  ENSEMBL        exon  13403.0  13655.0  .  +  .   \n",
       "13  chr1  ENSEMBL        exon  13661.0  14409.0  .  +  .   \n",
       "14  chr1   HAVANA  transcript  12010.0  13670.0  .  +  .   \n",
       "\n",
       "                                                    8                    ID  \\\n",
       "0   ID=ENSG00000223972.4;gene_id=ENSG00000223972.4...       ENSG00000223972   \n",
       "1   ID=ENST00000456328.2;Parent=ENSG00000223972.4;...       ENST00000456328   \n",
       "2   ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  exon:ENST00000456328   \n",
       "3   ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  exon:ENST00000456328   \n",
       "4   ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  exon:ENST00000456328   \n",
       "5   ID=ENST00000515242.2;Parent=ENSG00000223972.4;...       ENST00000515242   \n",
       "6   ID=exon:ENST00000515242.2:1;Parent=ENST0000051...  exon:ENST00000515242   \n",
       "7   ID=exon:ENST00000515242.2:2;Parent=ENST0000051...  exon:ENST00000515242   \n",
       "8   ID=exon:ENST00000515242.2:3;Parent=ENST0000051...  exon:ENST00000515242   \n",
       "9   ID=ENST00000518655.2;Parent=ENSG00000223972.4;...       ENST00000518655   \n",
       "10  ID=exon:ENST00000518655.2:1;Parent=ENST0000051...  exon:ENST00000518655   \n",
       "11  ID=exon:ENST00000518655.2:2;Parent=ENST0000051...  exon:ENST00000518655   \n",
       "12  ID=exon:ENST00000518655.2:3;Parent=ENST0000051...  exon:ENST00000518655   \n",
       "13  ID=exon:ENST00000518655.2:4;Parent=ENST0000051...  exon:ENST00000518655   \n",
       "14  ID=ENST00000450305.2;Parent=ENSG00000223972.4;...       ENST00000450305   \n",
       "\n",
       "             Parent  ExonLength  \n",
       "0               nan         0.0  \n",
       "1   ENSG00000223972         0.0  \n",
       "2   ENST00000456328       358.0  \n",
       "3   ENST00000456328       108.0  \n",
       "4   ENST00000456328      1188.0  \n",
       "5   ENSG00000223972         0.0  \n",
       "6   ENST00000515242       355.0  \n",
       "7   ENST00000515242       108.0  \n",
       "8   ENST00000515242      1187.0  \n",
       "9   ENSG00000223972         0.0  \n",
       "10  ENST00000518655       353.0  \n",
       "11  ENST00000518655       126.0  \n",
       "12  ENST00000518655       252.0  \n",
       "13  ENST00000518655       748.0  \n",
       "14  ENSG00000223972         0.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df_mod.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1196293, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>ExonLength</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>358.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon</td>\n",
       "      <td>exon:ENST00000456328</td>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>1188.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>exon</td>\n",
       "      <td>exon:ENST00000515242</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>355.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exon</td>\n",
       "      <td>exon:ENST00000515242</td>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      2                    ID           Parent  ExonLength\n",
       "2  exon  exon:ENST00000456328  ENST00000456328       358.0\n",
       "3  exon  exon:ENST00000456328  ENST00000456328       108.0\n",
       "4  exon  exon:ENST00000456328  ENST00000456328      1188.0\n",
       "6  exon  exon:ENST00000515242  ENST00000515242       355.0\n",
       "7  exon  exon:ENST00000515242  ENST00000515242       108.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Original: 2.6 M rows\n",
    "# Exon: 1.2M rows\n",
    "data_df_ExonLength = data_df_mod.loc[data_df_mod[2]=='exon',[2,'ID','Parent','ExonLength']]\n",
    "print(data_df_ExonLength.shape)\n",
    "data_df_ExonLength.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Parent\n",
       "ENST00000000233    1097.0\n",
       "ENST00000000412    2749.0\n",
       "ENST00000000442    2208.0\n",
       "ENST00000001008    2257.0\n",
       "ENST00000001146    4726.0\n",
       "Name: ExonLength, dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe that contains name of transcripts and length: data_df_TranscriptLength\n",
    "\n",
    "# Groupby Parent: 117592 rows\n",
    "data_df_TranscriptLength = data_df_ExonLength.groupby('Parent')['ExonLength'].sum()\n",
    "data_df_TranscriptLength.head(5)\n",
    "# data_df_TranscriptLength.shape\n",
    "# data_df_TranscriptLength.loc['ENST00000450305']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENST00000438504</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         transcript             gene\n",
       "1   ENST00000456328  ENSG00000223972\n",
       "5   ENST00000515242  ENSG00000223972\n",
       "9   ENST00000518655  ENSG00000223972\n",
       "14  ENST00000450305  ENSG00000223972\n",
       "22  ENST00000438504  ENSG00000227232"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe that contains name of gens and transcripts\n",
    "data_df_gen = data_df_mod.loc[data_df_mod[2]=='transcript',['ID','Parent']]\n",
    "data_df_gen = data_df_gen.rename(columns = {'ID':'transcript','Parent':'gene'})\n",
    "data_df_gen.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000456328</th>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000515242</th>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000518655</th>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305</th>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000438504</th>\n",
       "      <td>ENST00000438504</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      transcript             gene\n",
       "transcript                                       \n",
       "ENST00000456328  ENST00000456328  ENSG00000223972\n",
       "ENST00000515242  ENST00000515242  ENSG00000223972\n",
       "ENST00000518655  ENST00000518655  ENSG00000223972\n",
       "ENST00000450305  ENST00000450305  ENSG00000223972\n",
       "ENST00000438504  ENST00000438504  ENSG00000227232"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df_gen.index = data_df_gen['transcript']\n",
    "data_df_gen.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(196520, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcriptLength</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000456328</th>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>1654.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000515242</th>\n",
       "      <td>ENST00000515242</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>1650.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000518655</th>\n",
       "      <td>ENST00000518655</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>1479.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305</th>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000438504</th>\n",
       "      <td>ENST00000438504</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>1771.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      transcript             gene  transcriptLength\n",
       "ENST00000456328  ENST00000456328  ENSG00000223972            1654.0\n",
       "ENST00000515242  ENST00000515242  ENSG00000223972            1650.0\n",
       "ENST00000518655  ENST00000518655  ENSG00000223972            1479.0\n",
       "ENST00000450305  ENST00000450305  ENSG00000223972             626.0\n",
       "ENST00000438504  ENST00000438504  ENSG00000227232            1771.0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_total = pd.merge(data_df_gen, data_df_TranscriptLength, left_index=True, right_index=True, how='inner')\n",
    "df_total = df_total.rename(columns = {'ExonLength':'transcriptLength'})\n",
    "print(df_total.shape)\n",
    "df_total.head(5)\n",
    "# df_total.loc['ENST00000450305']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>transcriptLength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000264819</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000265350</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000265658</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000266731</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSGR0000270726</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57820 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 transcript  transcriptLength\n",
       "gene                                         \n",
       "ENSG00000000003           3                 3\n",
       "ENSG00000000005           2                 2\n",
       "ENSG00000000419           7                 7\n",
       "ENSG00000000457           5                 5\n",
       "ENSG00000000460          10                10\n",
       "...                     ...               ...\n",
       "ENSGR0000264819           1                 1\n",
       "ENSGR0000265350           1                 1\n",
       "ENSGR0000265658           1                 1\n",
       "ENSGR0000266731           1                 1\n",
       "ENSGR0000270726           2                 2\n",
       "\n",
       "[57820 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many transcripts per gene\n",
    "df_total.groupby('gene').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save file\n",
    "file='GFF3.csv'\n",
    "df_total.to_csv(file, sep=';', na_rep='n.a.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Desarrollo-OyWcDdAp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4 (tags/v3.10.4:9d38120, Mar 23 2022, 23:13:41) [MSC v.1929 64 bit (AMD64)]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "0c86d74f942837168ba3eb0ab04c2d3bf4a14a75d7820611aca17ecb14925afc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
